src/core/featuremodel.cpp

```diff
@@ -452,37 +454,61 @@ void FeatureModel::resetAttributes()
 
 void FeatureModel::applyGeometry()
 {
+  QString error;
   QgsGeometry geometry = mGeometry->asQgsGeometry();
 
-  QList<QgsVectorLayer *> intersectionLayers;
-
-  switch ( QgsProject::instance()->avoidIntersectionsMode() )
-  {
-    case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsCurrentLayer:
-      intersectionLayers.append( mLayer );
-      break;
-    case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsLayers:
-      intersectionLayers = QgsProject::instance()->avoidIntersectionsLayers();
-      break;
-    case QgsProject::AvoidIntersectionsMode::AllowIntersections:
-      break;
-  }
-  if ( !intersectionLayers.isEmpty() && geometry.type() == QgsWkbTypes::PolygonGeometry )
+  if ( QgsWkbTypes::geometryType( geometry.wkbType() ) == QgsWkbTypes::PolygonGeometry )
   {
-    geometry.avoidIntersections( intersectionLayers );
+    // Remove any invalid intersection in polygon geometry
+    QgsGeometry sanitizedGeometry;
+    if ( QgsGeometryCollection *collection = qgsgeometry_cast<QgsGeometryCollection *>( geometry.get() ) )
+    {
+      QgsGeometryPartIterator parts = collection->parts();
+      while ( parts.hasNext() )
+      {
+        QgsGeometry part( parts.next() );
+        sanitizedGeometry.addPart( part.buffer( 0.0, 5 ).constGet()->clone(), QgsWkbTypes::PolygonGeometry );
+      }
+    }
+    else if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( geometry.get() ) )
+    {
+      sanitizedGeometry = geometry.buffer( 0, 5 );
+    }
+    if ( sanitizedGeometry.constGet()->isValid( error ) )
+      geometry = sanitizedGeometry;
+
+    QList<QgsVectorLayer *> intersectionLayers;
+    switch ( QgsProject::instance()->avoidIntersectionsMode() )
+    {
+      case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsCurrentLayer:
+        intersectionLayers.append( mLayer );
+        break;
+      case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsLayers:
+        intersectionLayers = QgsProject::instance()->avoidIntersectionsLayers();
+        break;
+      case QgsProject::AvoidIntersectionsMode::AllowIntersections:
+        break;
+    }
+    if ( !intersectionLayers.isEmpty() )
+      geometry.avoidIntersections( intersectionLayers );
   }
 
   if ( mLayer && mLayer->geometryOptions()->geometryPrecision() != 0.0 )
   {
     const double precision = mLayer->geometryOptions()->geometryPrecision();
     QgsGeometry snappedGeometry = geometry.snappedToGrid( precision, precision );
-    geometry = snappedGeometry;
+    if ( snappedGeometry.constGet()->isValid( error ) )
```

Comment:
If geometryoptions are set, this will be done automatically when you add the feature to the layer. See https://github.com/qgis/QGIS/blob/master/src/core/qgsgeometryoptions.cpp#L56